drape/vulkan/vulkan_layers.cpp

```diff
@@ -0,0 +1,456 @@
+    return "DEBUG_REPORT_CALLBACK_EXT";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_DISPLAY_KHR_EXT: return "DISPLAY_KHR";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_DISPLAY_MODE_KHR_EXT: return "DISPLAY_MODE_KHR";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_OBJECT_TABLE_NVX_EXT: return "OBJECT_TABLE_NVX";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_INDIRECT_COMMANDS_LAYOUT_NVX_EXT:
+    return "INDIRECT_COMMANDS_LAYOUT_NVX";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_VALIDATION_CACHE_EXT_EXT: return "VALIDATION_CACHE_EXT";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_SAMPLER_YCBCR_CONVERSION_EXT: return "SAMPLER_YCBCR_CONVERSION";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_DESCRIPTOR_UPDATE_TEMPLATE_EXT:
+    return "DESCRIPTOR_UPDATE_TEMPLATE";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_ACCELERATION_STRUCTURE_NV_EXT:
+    return "ACCELERATION_STRUCTURE_NV";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_RANGE_SIZE_EXT: return "RANGE_SIZE";
+  case VK_DEBUG_REPORT_OBJECT_TYPE_MAX_ENUM_EXT: return "MAX_ENUM";
+  }
+  UNREACHABLE();
+  return {};
+}
+
+bool HasLayerOrExtension(char const * name, std::vector<char const *> const & collection)
+{
+  return collection.end() != std::find_if(collection.begin(), collection.end(),
+                                          [name](char const * v) { return strcmp(name, v) == 0; });
+}
+}  // namespace
+
+VkBool32 VKAPI_PTR DebugReportCallbackImpl(VkDebugReportFlagsEXT flags,
+                                           VkDebugReportObjectTypeEXT objectType, uint64_t object,
+                                           size_t location, int32_t messageCode,
+                                           const char * pLayerPrefix, const char * pMessage,
+                                           void * pUserData)
+{
+  if (flags & VK_DEBUG_REPORT_INFORMATION_BIT_EXT)
+  {
+    LOG(LINFO, ("Vulkan Diagnostics [", pLayerPrefix, "] [", GetReportObjectTypeString(objectType),
+                "] [OBJ:", object, "LOC:", location, "]:", pMessage));
+  }
+
+  if (flags & VK_DEBUG_REPORT_WARNING_BIT_EXT)
+  {
+    LOG(LWARNING, ("Vulkan Diagnostics [", pLayerPrefix, "] [", GetReportObjectTypeString(objectType),
+                   "] [OBJ:", object, "LOC:", location, "]:", pMessage));
+  }
+
+  if (flags & VK_DEBUG_REPORT_ERROR_BIT_EXT)
+  {
+    LOG(LERROR, ("Vulkan Diagnostics [", pLayerPrefix, "] [", GetReportObjectTypeString(objectType),
+                 "] [OBJ:", object, "LOC:", location, "]:", pMessage));
+  }
+
+  if (flags & VK_DEBUG_REPORT_DEBUG_BIT_EXT)
+  {
+    LOG(LDEBUG, ("Vulkan Diagnostics [", pLayerPrefix, "] [", GetReportObjectTypeString(objectType),
+                 "] [OBJ:", object, "LOC:", location, "]:", pMessage));
+  }
+
+  if (flags & VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT)
+  {
+    LOG(LWARNING, ("Vulkan Diagnostics (Performance) [", pLayerPrefix, "] [",
+                   GetReportObjectTypeString(objectType), "] [OBJ:", object, "LOC:", location,
+                   "]:", pMessage));
+  }
+
+  return VK_FALSE;
```

Comment:
всегда возвращаем VK_FALSE?

Reply:
Ага, так функция требует

Reply:
https://vulkan.lunarg.com/doc/view/1.0.37.0/linux/vkspec.chunked/ch32s02.html
"Applications should always return VK_FALSE so that they see the same behavior with and without validation layers enabled. If the application returns VK_TRUE from its callback and the Vulkan call being aborted returns a VkResult, the layer will return VK_ERROR_VALIDATION_FAILED_EXT."

Reply:
интересно, спасибо